File: tasks/python/mid/lru-ttl-cache/solution/ttlcache.py
"""LRU + TTL cache."""

import time
from collections import OrderedDict

_MISSING = object()


class TTLCache:
    def __init__(self, maxsize, ttl, clock=time.monotonic):
        if maxsize < 1:
            raise ValueError("maxsize must be >= 1")
        if ttl <= 0:
            raise ValueError("ttl must be > 0")
        self.maxsize = maxsize
        self.ttl = ttl
        self._clock = clock
        # key -> (value, stored_at); order = least recently used first
        self._data = OrderedDict()

    def _expired(self, stored_at, now=None):
        if now is None:
            now = self._clock()
        return now - stored_at >= self.ttl

    def _purge(self):
        now = self._clock()
        dead = [k for k, (_, t) in self._data.items() if self._expired(t, now)]
        for k in dead:
            del self._data[k]
        return len(dead)

    def set(self, key, value):
        if key in self._data:
            del self._data[key]
        else:
            if len(self._data) >= self.maxsize:
                self._purge()
            if len(self._data) >= self.maxsize:
                self._data.popitem(last=False)  # evict LRU
        self._data[key] = (value, self._clock())

    def get(self, key, default=None):
        entry = self._data.get(key)
        if entry is None:
            return default
        value, stored_at = entry
        if self._expired(stored_at):
            del self._data[key]
            return default
        self._data.move_to_end(key)  # most recently used
        return value

    def __contains__(self, key):
        entry = self._data.get(key)
        if entry is None:
            return False
        if self._expired(entry[1]):
            del self._data[key]
            return False
        return True

    def pop(self, key, default=None):
        entry = self._data.pop(key, _MISSING)
        if entry is _MISSING:
            return default
        value, stored_at = entry
        if self._expired(stored_at):
            return default
        return value

    def expire(self):
        return self._purge()

    def __len__(self):
        self._purge()
        return len(self._data)

    def keys(self):
        self._purge()
        return list(reversed(self._data.keys()))

File: tasks/dart/mid/lru-ttl-cache/solution/lib/ttl_cache.dart
/// Capacity-bounded cache with per-entry time-to-live.

class _Entry<V> {
  final V value;
  final int stamp;

  _Entry(this.value, this.stamp);
}

/// LRU cache whose entries expire [ttlMs] after their last write.
///
/// The [clock] returns the current time in milliseconds; injecting it keeps
/// eviction behavior fully deterministic for callers.
class TtlCache<K, V> {
  final int capacity;
  final int ttlMs;
  final int Function() clock;
  final Map<K, _Entry<V>> _map = {};

  TtlCache({required this.capacity, required this.ttlMs, required this.clock}) {
    if (capacity < 1) throw ArgumentError('capacity must be >= 1');
    if (ttlMs < 1) throw ArgumentError('ttlMs must be >= 1');
  }

  bool _expired(_Entry<V> entry) => clock() - entry.stamp >= ttlMs;

  void _prune() {
    _map.removeWhere((_, entry) => _expired(entry));
  }

  V? get(K key) {
    final entry = _map[key];
    if (entry == null) return null;
    if (_expired(entry)) {
      _map.remove(key);
      return null;
    }
    // Refresh recency: reinsertion moves the key to the back of the
    // insertion order, which is the most-recently-used position.
    _map.remove(key);
    _map[key] = entry;
    return entry.value;
  }

  void put(K key, V value) {
    final now = clock();
    if (_map.containsKey(key)) {
      // Overwrite: never evicts, just restamp and refresh recency.
      _map.remove(key);
    } else {
      _prune();
      if (_map.length >= capacity) {
        _map.remove(_map.keys.first);
      }
    }
    _map[key] = _Entry(value, now);
  }

  bool containsKey(K key) {
    final entry = _map[key];
    if (entry == null) return false;
    if (_expired(entry)) {
      _map.remove(key);
      return false;
    }
    return true;
  }

  void remove(K key) {
    _map.remove(key);
  }

  int get length {
    _prune();
    return _map.length;
  }

  List<K> get keys {
    _prune();
    return List.of(_map.keys);
  }
}
